operators/src/source/gdal_source/mod.rs

```diff
@@ -588,33 +695,36 @@ impl InitializedRasterOperator for InitializedGdalSourceOperator {
 /// This method reads the data for a single grid with a specified size from the GDAL dataset.
 /// It fails if the tile is not within the dataset.
 #[allow(clippy::float_cmp)]
-fn read_grid_from_raster<
-    T,
-    D: GridSize<ShapeArray = [usize; 2]> + GridSpaceToLinearSpace<IndexArray = [isize; 2]>,
->(
+fn read_grid_from_raster<T, D>(
     dataset: &gdal::Dataset,
-    dataset_grid_box: &GridBoundingBox2D,
-    tile_grid: D,
+    read_window: &GdalReadWindow,
+    out_shape: D,
     dataset_params: &GdalDatasetParameters,
+    flip_y_axis: bool,
 ) -> Result<GridOrEmpty<D, T>>
 where
     T: Pixel + GdalType + Default + FromPrimitive,
-    D: PartialEq + Clone,
+    D: Clone + GridSize + PartialEq,
 {
     let raster_band_index = dataset_params.rasterband_channel;
 
     let rasterband = dataset.rasterband(raster_band_index as isize)?;
 
-    let GridIdx([dataset_ul_y, dataset_ul_x]) = dataset_grid_box.min_index();
-    let [dataset_y_size, dataset_x_size] = dataset_grid_box.axis_size();
-    let [tile_y_size, tile_x_size] = tile_grid.axis_size();
+    let gdal_out_shape = (out_shape.axis_size_x(), out_shape.axis_size_y());
+
     let buffer = rasterband.read_as::<T>(
-        (dataset_ul_x, dataset_ul_y),     // pixelspace origin
-        (dataset_x_size, dataset_y_size), // pixelspace size
-        (tile_x_size, tile_y_size),       // requested raster size
-        None,                             // sampling mode
+        read_window.gdal_window_start(), // pixelspace origin
+        read_window.gdal_window_size(),  // pixelspace size
+        gdal_out_shape,                  // requested raster size
+        None,                            // sampling mode
     )?;
-    let data_grid = Grid::new(tile_grid.clone(), buffer.data)?;
+    let data_grid = Grid::new(out_shape.clone(), buffer.data)?;
+
+    let data_grid = if flip_y_axis {
```

Comment:
would it make sense to do this outside of this function, and only do the actual reading here? Also: maybe implement the reverse on the MaskedGrid so that it only has to be called once?

Reply:
i thought about this and if we do that we need to reverse the mask also. even if it was generated from no-data value or if the data is always valid.